Support hiding columns

kyleam opened the pull request:
pyout requires that all columns are specified before writing begins.
If none are explicitly specified, it infers the list of columns from
the first record received.

For some uses cases, it'd be desirable to expand the list of columns
as new keys are encountered (gh-49).  Unfortunately many parts of
pyout assume that the set of columns is fixed, so adding support for
auto-expanding the column list seems like it will require pretty deep
changes.  The issue is complicated by the fact that the request
involves more than being able to give records with new keys when
calling Tabular: asynchronous callbacks, themselves specified as
values to some column(s), should also be able to create new columns.

One of the motivations for auto-expanding the column list is the
desire to not show columns unless they actually contain information.
Address this by allowing columns to be hidden until they contain
something other than a Nothing() value.  Because it's easy to do and
could end up being useful (at least for debugging), also allow columns
to be unconditionally hidden.

The other motivation for auto-expanding columns is to let the caller
avoid specifying (or perhaps even knowing) the column names up front.
Hiding of course doesn't help with this.

Re: #49, #84

yarikoptic commented:
THANK YOU!  I will change dandi interaction with pyout a bit and try it out!

yarikoptic commented:
Woohoo, thanks!  It works, but there is a few gotchas with formatting.  

<details>
<summary>E.g. on https://github.com/dandi-datasets/nwb_test_data with dandi-cli (04d4d873725865e480e6781c4653ab9926a8971a) I am getting following
</summary> 

```shell
$> dandi ls /tmp/nwb_test_data/v2.0.1/*
PATH                                                          SIZE     N... NWB  SESSION_DESCRIPTION                                             NUMBER_OF_UNITS IDENTIFIER                      SESSION_START_TIME
/tmp/nwb_test_data/v2.0.1/test_Clustering.nwb                 17.8 kB       2.0b a file to test writing and reading a Clustering                                 TEST_Clustering                 1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_ClusterWaveforms.nwb           20.0 kB       2.0b a file to test writing and reading a ClusterWaveforms                           TEST_ClusterWaveforms           1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_CurrentClampSeries.nwb         25.5 kB       2.0b a file to test writing and reading a CurrentClampSeries                         TEST_CurrentClampSeries         1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_CurrentClampStimulusSeries.nwb 24.7 kB       2.0b a file to test writing and reading a CurrentClampStimulusSeries                 TEST_CurrentClampStimulusSeries 1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_DecompositionSeries.nwb        24.8 kB       2.0b a file to test writing and reading a DecompositionSeries                        TEST_DecompositionSeries        1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_Device.nwb                     17.4 kB       2.0b a file to test writing and reading a Device                                     TEST_Device                     1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_DynamicTable.nwb               18.0 kB       2.0b a file to test writing and reading a DynamicTable               2               TEST_DynamicTable               1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_ElectricalSeries.nwb           29.5 kB  4    2.0b a file to test writing and reading a ElectricalSeries                           TEST_ElectricalSeries           1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_ElectrodeGroup.nwb             19.8 kB       2.0b a file to test writing and reading a ElectrodeGroup                             TEST_ElectrodeGroup             1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_EventDetection.nwb             31.7 kB  4    2.0b a file to test writing and reading a EventDetection                             TEST_EventDetection             1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_EventWaveform.nwb              30.7 kB  4    2.0b a file to test writing and reading a EventWaveform                              TEST_EventWaveform              1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_FeatureExtraction.nwb          29.3 kB  4    2.0b a file to test writing and reading a FeatureExtraction                          TEST_FeatureExtraction          1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_FilteredEphys.nwb              33.6 kB  4    2.0b a file to test writing and reading a FilteredEphys                              TEST_FilteredEphys              1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_ImagingPlane.nwb               23.3 kB       2.0b a file to test writing and reading a ImagingPlane                               TEST_ImagingPlane               1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_IntracellularElectrode.nwb     21.7 kB       2.0b a file to test writing and reading a IntracellularElectrode                     TEST_IntracellularElectrode     1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_IZeroClampSeries.nwb           25.5 kB       2.0b a file to test writing and reading a IZeroClampSeries                           TEST_IZeroClampSeries           1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_LFP.nwb                        33.6 kB  4    2.0b a file to test writing and reading a LFP                                        TEST_LFP                        1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_OptogeneticSeries.nwb          23.2 kB       2.0b a file to test writing and reading a OptogeneticSeries                          TEST_OptogeneticSeries          1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_OptogeneticStimulusSite.nwb    20.6 kB       2.0b a file to test writing and reading a OptogeneticStimulusSite                    TEST_OptogeneticStimulusSite    1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_PatchClampSeries.nwb           28.2 kB        2.0b a file to test writing and reading a PatchClampSeries                             TEST_PatchClampSeries             1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_PlaneSegmentation.nwb          34.4 kB        2.0b a file to test writing and reading a PlaneSegmentation                            TEST_PlaneSegmentation            1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_RoiResponseSeries.nwb          37.6 kB        2.0b a file to test writing and reading a RoiResponseSeries                            TEST_RoiResponseSeries            1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_Subject.nwb                    18.8 kB        2.0b a file to test writing and reading a Subject                                      TEST_Subject                      1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_SweepTable.nwb                 28.2 kB        2.0b a file to test writing and reading a SweepTable                                   TEST_SweepTable                   1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_TimeIntervals.nwb              27.2 kB        2.0b a file to test writing and reading a TimeIntervals                                TEST_TimeIntervals                1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_TimeSeries.nwb                 17.6 kB        2.0b a file to test writing and reading a TimeSeries                                   TEST_TimeSeries                   1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_timestamps_linking.nwb         43.5 kB        2.0b bar                                                                               foo                               2019-03-21/19:13:50
/tmp/nwb_test_data/v2.0.1/test_TwoPhotonSeries.nwb            26.9 kB        2.0b a file to test writing and reading a TwoPhotonSeries                              TEST_TwoPhotonSeries              1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_Units.nwb                      19.3 kB        2.0b a file to test writing and reading a Units                                        TEST_Units                        1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_VoltageClampSeries.nwb         27.1 kB        2.0b a file to test writing and reading a VoltageClampSeries                           TEST_VoltageClampSeries           1971-01-01/12:00:00
/tmp/nwb_test_data/v2.0.1/test_VoltageClampStimulusSeries.nwb 24.7 kB        2.0b a file to test writing and reading a VoltageClampStimulusSeries                   TEST_VoltageClampStimulusSeries   1971-01-01/12:00:00
Summary:                                                      804.1 kB                                                                                                                                1971-01-01/12:00:00>
                                                                                                                                                                                                      2019-03-21/19:13:50<
```
</details>

- with `/tmp/nwb_test_data/v2.0.1/test_PlaneSegmentation.nwb`  fields starting from `SESSION_START_TIME` get indented more than needed

- 3rd column in my case (`number_of_electrodes`) has only `N...` in the header.  I will shorten those "number_of" names, but it seems to point that pyout seems to fail to fairly split widths (some other fields aren't truncated at all) or smth like that.

Those above are more likely to be related to this PR than not, since if I specify columns explicitly
<details>
<summary>with the following diff (yet to formalize into proper cmdline arg):</summary> 

```shell
$> git diff
diff --git a/dandi/cli/command.py b/dandi/cli/command.py
index 2f353ba..b0f6bdf 100644
--- a/dandi/cli/command.py
+++ b/dandi/cli/command.py
@@ -191,8 +191,11 @@ def ls(paths):
         # a tty
         PYOUT_STYLE["width_"] = 200
 
+    columns = ['path', 'size', 'session_start_time', 'nwb_version', 'identifier',
+               'session_description', 'number_of_electrodes', 'number_of_units']
     out = pyout.Tabular(
         # TODO: provide columns=[] from cmdline arg
+        columns=columns,
         style=PYOUT_STYLE
         # , stream=...
     )
@@ -200,7 +203,7 @@ def ls(paths):
         from ..pynwb_utils import metadata_all_fields
 
         # TODO: more logical ordering
-        async_keys = tuple(set(metadata_all_fields))
+        async_keys = tuple(set(metadata_all_fields).intersection(columns))
         for path in files:
             rec = {"path": path}
             try:
```
</details>
then it looks ok.

kyleam commented:
> It works, but there is a few gotchas with formatting.
>
> * with `/tmp/nwb_test_data/v2.0.1/test_PlaneSegmentation.nwb`  fields starting from `SESSION_START_TIME` get indented more than needed

The issue here is that, even if there's no width left for an updated field to claim, the flanking separators were still added for the updated row.  Addressed in 6fd8401d.

> * 3rd column in my case (`number_of_electrodes`) has only `N...` in the header.  I will shorten those "number_of" names, but it seems to point that pyout seems to fail to fairly split widths (some other fields aren't truncated at all) or smth like that.

pyout doesn't yet try to do much at all to "fairly" split widths.  The most it does is sort the fields by increasing width so that smaller fields have a chance to claim width before the large ones claim it.  Once all the free width is claimed, that's it.  Without hiding, that's not ideal if you're relying on auto-width for all the columns and have a lot of callbacks return long values later.  With hiding, it's very problematic because it can hide entire columns.  That will need to be sorted out before merging this in.

kyleam commented:
I've now rebased this onto the gh-90, which should result in widths being divided more evenly across columns.

Here's the [restricted diff][0] for this PR.

[0]: https://github.com/pyout/pyout/compare/0b0ea6d06d84ce12b2dfccac0613da29f87fce88...kyleam:hide-columns

<details>
<summary>range-diff</summary>

```
1:  6fd8401d ! 1:  064a3f79 common: Don't render zero-width fields
    @@ pyout/tests/test_tabular.py: def test_tabular_write_empty_string():
      
      
      def test_tabular_write_missing_column_missing_text():
    -@@ pyout/tests/test_tabular.py: def test_tabular_auto_width_exceeds_total():
    -     out(OrderedDict([("name", "foobert"),
    -                      ("status", "okiguess"),
    -                      ("path", "illbedropped:(")]))
    --    assert out.stdout == "foobert o... \n"
    -+    assert out.stdout == "foobert o...\n"
    - 
    - 
    - def test_tabular_auto_width_exceeds_total_multiline():
     @@ pyout/tests/test_tabular.py: def test_tabular_write_callable_transform_nothing():
              # function above, which is designed to take a number, won't be called
              # with it.
2:  83b528da = 2:  80cb7b5d Rename test to avoid word "hidden"
-:  -------- > 3:  63233254 common: Move logic for setting fixed widths to a method
-:  -------- > 4:  21754e33 common: Cosmetic tweaks to _set_fixed_widths()
3:  262da323 ! 5:  ae1c6bb5 summary: Explicitly specify columns to summarize
    @@ Commit message
         caller to filter out the hidden columns.
     
      ## pyout/common.py ##
    -@@ pyout/common.py: def init_columns(self, columns, ids):
    -     def update(self, row, style):
    +@@ pyout/common.py: def update(self, row, style):
    +         lgr.log(9, "Updating with .summary set to %s", self.summary)
              content, status = super(ContentWithSummary, self).update(row, style)
              if self.summary:
     -            summ_rows = self.summary.summarize([r.row for r in self._rows])
4:  edd8536b ! 6:  9218ffce Support hiding columns
    @@ Commit message
         avoid specifying (or perhaps even knowing) the column names up front.
         Hiding of course doesn't help with this.
     
    +    Note that are a couple of details that make this a little more
    +    complicated than simply skipping the hidden fields when rendering:
    +
    +      * The header and summary should not be able to trigger unhiding.
    +
    +      * After unhiding a column we need to reset fixed values that were
    +        calculated in build()
    +
         Re: #49, #84
     
      ## CHANGELOG.md ##
    @@ CHANGELOG.md: and this project adheres to [Semantic Versioning](http://semver.or
      
      ## [Unreleased]
      
    ++### Added
    ++
     +- A new style attribute, "hide", makes it possible to hide a column,
     +  either unconditionally or until `Tabular` is called with a record
     +  that includes the column.
      
    - ## [0.4.1] - 2019-10-02
    + ### Changed
      
     
      ## pyout/common.py ##
    @@ pyout/common.py: def __init__(self, style, procgen):
              self._truncaters = {}
      
     +        self.hidden = {}  # column => {True, "if-empty", False}
    ++        self._visible_columns = None  # cached list of visible columns
     +
          def build(self, columns):
              """Build the style and fields.
      
     @@ pyout/common.py: def build(self, columns):
    +             "width_", elements.default("width_"))
              elements.validate(self.style)
              self._setup_fields()
    +-        self._set_fixed_widths()
      
    +-        self._check_widths()
     +        self.hidden = {c: self.style[c]["hide"] for c in columns}
    ++        self._reset_width_info()
    + 
    +     def _compose(self, name, attributes):
    +         """Construct a style taking `attributes` from the column styles.
    +@@ pyout/common.py: def has_header(self):
    +         """
    +         return self.style["header_"] is not None
    + 
    ++    @property
    ++    def visible_columns(self):
    ++        """List of columns that are not marked as hidden.
    ++
    ++        This value is cached and becomes invalid if column visibility has
    ++        changed since the last `render` call.
    ++        """
    ++        if self._visible_columns is None:
    ++            hidden = self.hidden
    ++            self._visible_columns = [c for c in self.columns if not hidden[c]]
    ++        return self._visible_columns
     +
    -         ngaps = len(self.columns) - 1
    -         self.width_separtor = len(self.style["separator_"]) * ngaps
    -         lgr.debug("Calculated separator width as %d", self.width_separtor)
    +     def _check_widths(self):
    +-        columns = self.columns
    ++        visible = self.visible_columns
    +         autowidth_columns = self.autowidth_columns
    +         width_table = self.style["width_"]
    +-        if len(columns) > width_table:
    ++        if len(visible) > width_table:
    +             raise elements.StyleError(
    +-                "Number of columns exceeds available table width")
    ++                "Number of visible columns exceeds available table width")
    + 
    +         width_fixed = self.width_fixed
    +         width_auto = width_table - width_fixed
    + 
    +-        if width_auto < len(autowidth_columns):
    ++        if width_auto < len(set(autowidth_columns).intersection(visible)):
    +             raise elements.StyleError(
    +-                "The number of auto-width columns ({}) "
    ++                "The number of visible auto-width columns ({}) "
    +                 "exceeds the available width ({})"
    +                 .format(len(autowidth_columns), width_auto))
    + 
    +     def _set_fixed_widths(self):
    +         """Set fixed-width attributes.
    ++
    ++        Previously calculated values are invalid if the number of visible
    ++        columns changes.  Call _reset_width_info() in that case.
    +         """
    +-        columns = self.columns
    +-        ngaps = len(columns) - 1
    ++        visible = self.visible_columns
    ++        ngaps = len(visible) - 1
    +         width_separtor = len(self.style["separator_"]) * ngaps
    +         lgr.debug("Calculated separator width as %d", width_separtor)
    + 
    +         autowidth_columns = self.autowidth_columns
    +         fields = self.fields
    +-        width_fixed = sum([sum(fields[c].width for c in columns
    ++        width_fixed = sum([sum(fields[c].width for c in visible
    +                                if c not in autowidth_columns),
    +                            width_separtor])
    +         lgr.debug("Calculated fixed width as %d", width_fixed)
    +@@ pyout/common.py: def _set_fixed_widths(self):
    +         self.width_separtor = width_separtor
    +         self.width_fixed = width_fixed
    + 
    ++    def _reset_width_info(self):
    ++        """Reset visibility-dependent information.
    ++        """
    ++        self._visible_columns = None
    ++        self._set_fixed_widths()
    ++        self._check_widths()
    ++
    +     def _set_widths(self, row, proc_group):
    +         """Update auto-width Fields based on `row`.
    + 
     @@ pyout/common.py: def _set_widths(self, row, proc_group):
      
    +         # Check what width each row wants.
              lgr.debug("Checking width for row %r", row)
    -         adjusted = False
     +        hidden = self.hidden
    -         for column in sorted(self.columns, key=lambda c: self.fields[c].width):
    +         for column in autowidth_columns:
     +            if hidden[column]:
    -+                lgr.debug("Not adjusting width for hidden column %r",
    ++                lgr.debug("%r is hidden; setting width to 0",
     +                          column)
    ++                autowidth_columns[column]["wants"] = 0
     +                continue
    -             # ^ Sorting the columns by increasing widths isn't necessary; we do
    -             # it so that columns that already take up more of the screen don't
    -             # continue to grow and use up free width before smaller columns
    ++
    +             field = fields[column]
    +             lgr.debug("Checking width of column %r (current field width: %d)",
    +                       column, field.width)
     @@ pyout/common.py: def _proc_group(self, style, adopt=True):
              else:
                  return "default"
    @@ pyout/common.py: def render(self, row, style=None, adopt=True):
              the field widths required adjustment (bool).
              """
     +        hidden = self.hidden
    ++        any_unhidden = False
     +        if can_unhide:
     +            for c in row:
     +                val = row[c]
    @@ pyout/common.py: def render(self, row, style=None, adopt=True):
     +                    lgr.debug("Unhiding column %r after encountering %r",
     +                              c, val)
     +                    hidden[c] = False
    ++                    any_unhidden = True
    ++        if any_unhidden:
    ++            self._reset_width_info()
     +
              group = self._proc_group(style, adopt=adopt)
              if group == "override":
    @@ pyout/common.py: def render(self, row, style=None, adopt=True):
      
              adjusted = self._set_widths(row, group)
     -        cols = self.columns
    -+        cols = (c for c in self.columns if not hidden[c])
    ++        cols = self.visible_columns
              proc_fields = ((self.fields[c], row[c]) for c in cols)
              # Exclude fields that weren't able to claim any width to avoid
              # surrounding empty values with separators.
    @@ pyout/common.py: def _add_header(self):
                                              "adopt": False})
      
      
    -@@ pyout/common.py: def init_columns(self, columns, ids):
    -     def update(self, row, style):
    +@@ pyout/common.py: def update(self, row, style):
              content, status = super(ContentWithSummary, self).update(row, style)
              if self.summary:
    -+            hidden = self.fields.hidden
    -+            columns = [c for c in self.columns if not hidden[c]]
                  summ_rows = self.summary.summarize(
     -                self.columns,
    -+                columns,
    ++                self.fields.visible_columns,
                      [r.row for r in self._rows])
      
                  def join():
    @@ pyout/tests/test_tabular.py: def test_tabular_write_inspect_with_getitem():
     +    assert_contains_nc(lines1, "bar  ok     b     ")
     +    assert_contains_nc(lines1, "X    1            ")
     +
    ++
    ++def test_tabular_hidden_col_takes_back_auto_space():
    ++    out = Tabular(["name", "status", "letter"],
    ++                  style={"width_": 10,
    ++                         "default_": {"width": {"marker": "…"}},
    ++                         "status": {"hide": "if_missing"}})
    ++    out({"name": "foo", "letter": "abcdefg"})
    ++    assert out.stdout.splitlines() == ["foo abcde…"]
    ++    out({"name": "foo", "status": "ok"})
    ++    assert_contains_nc(out.stdout.splitlines(), "foo ok ab…")
    ++
     +
      def test_tabular_summary():
      
```

</details>

kyleam commented:
Updated to resolve conflicts with recently merged gh-97.  Based on running the tests locally, it looks like there is now a line missing coverage.  I'll have to look into it.

<details>
<summary>range-diff</summary>

```
1:  064a3f79 = 1:  0847f5b3 common: Don't render zero-width fields
2:  80cb7b5d ! 2:  d06253cf Rename test to avoid word "hidden"
    @@ pyout/tests/test_tabular.py: def test_tabular_write_callable_values_multi_return
      @pytest.mark.parametrize("nrows", [20, 21])
     -def test_tabular_callback_to_hidden_row(nrows):
     +def test_tabular_callback_to_offscreen_row(nrows):
    -     if sys.version_info < (3,):
    -         try:
    -             import jsonschema
    +     delay = Delayed("OK")
    +     out = Tabular(style={"status": {"aggregate": len}})
    +     with out:
3:  63233254 = 3:  80a8c18e common: Move logic for setting fixed widths to a method
4:  21754e33 = 4:  f1ea7880 common: Cosmetic tweaks to _set_fixed_widths()
5:  ae1c6bb5 ! 5:  cdedd081 summary: Explicitly specify columns to summarize
    @@ pyout/common.py: def update(self, row, style):
                      return "".join(self._render(summ_rows))
     
      ## pyout/summary.py ##
    -@@ pyout/summary.py: def __bool__(self):
    - 
    -     __nonzero__ = __bool__  # py2
    +@@ pyout/summary.py: def __init__(self, style):
    +     def __bool__(self):
    +         return self._enabled
      
     -    def summarize(self, rows):
     -        """Return summary rows for `rows`.
6:  9218ffce ! 6:  f3642998 Support hiding columns
    @@ Commit message
     
      ## CHANGELOG.md ##
     @@ CHANGELOG.md: and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
    - 
    - ## [Unreleased]
    + - Support for Python 2 has been dropped.  Python 3.4 or later is
    +   required.
      
     +### Added
     +
```

</details>

kyleam commented:
> Based on running the tests locally, it looks like there is now a line missing coverage. I'll have to look into it.

Updated.
